torch/lib/TensorLib/gen.py

```diff
@@ -0,0 +1,190 @@
+import os
+import sys
+import yaml
+from optparse import OptionParser
+
+import cwrap_parser
+import preprocess_declarations
+import function_wrapper
+import dispatch_macros
+from code_template import CodeTemplate
+
+
+parser = OptionParser()
+parser.add_option('-s', '--source-path', help='path to source director for tensorlib',
+                  action='store', default='.')
+parser.add_option('-p', '--print-dependencies',
+                  help='only output a list of dependencies', action='store_true')
+parser.add_option('-n', '--no-cuda', action='store_true')
+options, cwrap_files = parser.parse_args()
+
+TEMPLATE_PATH = options.source_path + "/templates"
+GENERATOR_DERIVED = CodeTemplate.from_file(
+    TEMPLATE_PATH + "/GeneratorDerived.h")
+STORAGE_DERIVED_CPP = CodeTemplate.from_file(
+    TEMPLATE_PATH + "/StorageDerived.cpp")
+STORAGE_DERIVED_H = CodeTemplate.from_file(TEMPLATE_PATH + "/StorageDerived.h")
+
+TYPE_DERIVED_CPP = CodeTemplate.from_file(TEMPLATE_PATH + "/TypeDerived.cpp")
+TYPE_DERIVED_H = CodeTemplate.from_file(TEMPLATE_PATH + "/TypeDerived.h")
+TYPE_H = CodeTemplate.from_file(TEMPLATE_PATH + "/Type.h")
+TYPE_CPP = CodeTemplate.from_file(TEMPLATE_PATH + "/Type.cpp")
+
+TENSOR_DERIVED_CPP = CodeTemplate.from_file(
+    TEMPLATE_PATH + "/TensorDerived.cpp")
+TENSOR_DERIVED_H = CodeTemplate.from_file(TEMPLATE_PATH + "/TensorDerived.h")
+TENSOR_H = CodeTemplate.from_file(TEMPLATE_PATH + "/Tensor.h")
+
+FUNCTIONS_H = CodeTemplate.from_file(TEMPLATE_PATH + "/Functions.h")
+
+generators = {
+    'CPUGenerator.h': {
+        'name': 'CPU',
+        'th_generator': 'THGenerator * generator;',
+        'header': 'TH/TH.h',
+    },
+    'CUDAGenerator.h': {
+        'name': 'CUDA',
+        'th_generator': '',
+        'header': 'THC/THC.h'
+    },
+}
+
+processors = ['CPU']
+if not options.no_cuda:
+    processors.append('CUDA')
+
+scalar_types = [
+    ('Byte', 'uint8_t', 'Long'),
+    ('Char', 'int8_t', 'Long'),
+    ('Double', 'double', 'Double'),
+    ('Float', 'float', 'Double'),
+    ('Int', 'int', 'Long'),
+    ('Long', 'int64_t', 'Long'),
+    ('Short', 'int16_t', 'Long'),
+    ('Half', 'Half', 'Double'),
+]
+
+# shared environment for non-derived base classes Type.h Tensor.h Storage.h
+top_env = {
+    'type_registrations': [],
+    'type_headers': [],
+    'type_method_declarations': [],
+    'type_method_definitions': [],
+    'tensor_method_declarations': [],
+    'tensor_method_definitions': [],
+    'function_declarations': [],
+    'function_definitions': [],
+}
+
+
+def write(filename, s):
```

Comment:
These is needed for CMake. The first pass we generate the list of files, so that CMake knows about them. The second time we actually write the code.